### PR TITLE
Clean resolve inherited component

### PR DESCRIPTION
**Pull request by thk123**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [n/a] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [n/a] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by thk123** [comment]
Was looking at this to do the refactor of java_bytecode_convert_method, but the `resolve_inherited_componentt` relies on the globals already being present in the symbol table so isn't trivially applicable here, instead I will perform the tidy up requested.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5e74f0a43ab4df78fa3c81a206998fefe572a7c0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98344502

**Review by tautschnig — APPROVED** [review]
I really like negative diff stats :-)

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -3147,14 +3147,9 @@ bool java_bytecode_convert_methodt::is_method_inherited(
...
 {
-  resolve_inherited_componentt::inherited_componentt inherited_method =
-    get_inherited_component(
-      classname,
-      methodid,
-      symbol_table,
-      class_hierarchy,
-      false);
-  return inherited_method.is_valid();
+  const auto &inherited_method = get_inherited_component(
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Not a reference (0394j4)

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 2]
```diff
@@ -3166,18 +3161,13 @@ irep_idt java_bytecode_convert_methodt::get_static_field(
...
   const irep_idt &component_name) const
 {
-  resolve_inherited_componentt::inherited_componentt inherited_method =
-    get_inherited_component(
-      class_identifier,
-      component_name,
-      symbol_table,
-      class_hierarchy,
-      true);
+  const auto &inherited_method = get_inherited_component(
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: (0394j4)

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 3]
```diff
@@ -591,14 +591,9 @@ static void create_stub_global_symbols(
...
         // define static fields.
-        resolve_inherited_componentt::inherited_componentt referred_component =
-          get_inherited_component(
-            class_id,
-            component,
-            symbol_table,
-            class_hierarchy,
-            true);
-        if(!referred_component.is_valid())
+        const auto &referred_component = get_inherited_component(
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: (0394j4)

File: src/goto-programs/remove_virtual_functions.cpp [hunk 4]
```diff
@@ -399,18 +399,16 @@ void remove_virtual_functionst::get_child_functions_rec(
     }
     if(!function.symbol_expr.has_value())
     {
-      const resolve_inherited_componentt::inherited_componentt
-        &resolved_call = resolve_function_call(child, component_name);
-      if(resolved_call.is_valid())
+      const auto &resolved_call = resolve_function_call(child, component_name);
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: (0394j4)

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 5]
```diff
@@ -339,22 +340,22 @@ resolve_inherited_componentt::inherited_componentt get_inherited_component(
 {
   resolve_inherited_componentt component_resolver(
     symbol_table, class_hierarchy);
-  const resolve_inherited_componentt::inherited_componentt resolved_component =
+  const auto &resolved_component =
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_utils.cpp]: (0394j4)

File: src/goto-programs/remove_virtual_functions.cpp [hunk 6]
```diff
@@ -449,22 +447,22 @@ void remove_virtual_functionst::get_functions(
     return get_virtual_call_target(class_id, function_name, false);
     };
 
-  const resolve_inherited_componentt::inherited_componentt
-    &resolved_call = get_virtual_call_target(class_id, function_name, false);
+  const auto &resolved_call =
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: (0394j4)

**Review by smowton — COMMENTED** [review]
Don't refer to `optionalt` returned by value using a reference type -- this works due to automagical lifetime extension but gives the impression that it refers to something in-place when it's actually a local.

**Comment by smowton** [comment]
^ (0394j4) is like your emoji thing except I don't know how to type emoji :)

**Comment by thk123** [comment]
@smowton For future reference, type a `:` and then the name of the emoji :stuck_out_tongue:

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 1 above)
Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Initially thought it was C++ spec reference, and was very impressed!

**Comment by thk123** [comment]
@smowton addressed.

**Review by smowton — APPROVED** [review]

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 7]
```diff
@@ -548,11 +548,10 @@ irep_idt ci_lazy_methodst::get_virtual_method_target(
     return irep_idt();
 
   resolve_inherited_componentt call_resolver(symbol_table, class_hierarchy);
-  const resolve_inherited_componentt::inherited_componentt resolved_call =
-    call_resolver(classname, call_basename, false);
+  const auto &resolved_call = call_resolver(classname, call_basename, false);
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: 🚫 Shouldn't be a reference

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9c313499dbbf0bce049a574c0f67132884b1aa54).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98361182

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5bc48f83e8d172a1b914b165c043d894da458618).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98445489

**Pull request merged**